lib/henkei.rb

```diff
@@ -211,20 +219,34 @@ def self.java_path
 
   # Internal helper for calling to Tika library directly
   #
-  def self.client_read(type, data, include_ocr: false)
-    Open3.capture2(*tika_command(type, include_ocr: include_ocr), stdin_data: data, binmode: true).first
+  def self.client_read(type, data, include_ocr: false, encoding: nil)
+    unless encoding.nil? || Encoding.name_list.include?(encoding)
+      raise ArgumentError, "unsupported encoding - #{encoding}"
+    end
+
+    Open3.popen2(*tika_command(type, include_ocr: include_ocr, encoding: encoding)) do |stdin, stdout|
+      stdin.binmode
+      stdout.binmode
+      stdout.set_encoding encoding unless encoding.nil?
+
+      stdin.puts data
```

Comment:
`capture2` uses `write` at this stage. It is also wrapped in a begin/rescue block catching the `Errno::EPIPE` error.

Looking at the documentation/source they do slightly different things.. including `puts` writing an extra newline